version-upgrade-operator/README.md

```diff
@@ -0,0 +1,673 @@
+for application deployment (like helm-controller or app-operator).
+
+We design the version setting for App CRs and HelmReleases, but the solution is not CR specific. Using the new
+CRDs, an operator will be responsible for setting target CRs' selected property, like the `spec.version` of
+App CR, to an automatically determined value. The value will be discovered in a Helm charts repository and
+applied to a target CR only when time-based restrictions are met.
+
+### Image tags and registries discovery
+
+For the solution to work, we have to discover tags available for specific image (repository) in a registry. We
+figured out that instead of implementing our own logic for discovering tags, we can reuse the existing logic
+present in Flux controllers.
+
+#### Reusing Flux components
+
+Flux has version discovery implemented in a few places:
+
+- `HelmRelease/HelmRepository/HelmChart` level: these are the objects that drive actual Helm Chart deployment.
+  We want to edit them to point to a limited set of possible versions to make them work with the operator
+  we're designing
+- `OCIRepository` which allows to discover and filter tags of artifacts in an OCI repository; it is always
+  downloading the most recent matching artifact
+- `ImageRepository` and `ImagePolicy` are objects normally used by
+  [`ImageAutomationController`](https://fluxcd.io/flux/components/image/), but we can use them for easy
+  discovery of what's available in a registry.
+
+To not interfere with how upgrades might be already configured for existing `HelmReleases`, we will use a
+dedicated `ImageRepository` and `ImagePolicy` objects for discovery process related to automatic upgrades.
+
+##### How this would work?
+
+To configure an automated update for an App CR, we need answers to 3 key questions: 1) what should be
+upgraded; 2) when it should be upgraded; 3) which version should be used. The following steps have to be
+performed by a user and the new operator (this assumes that there's no GitOps interfering, deployment with
+GitOps is described below):
+
+1. The user deploys an application, either by using `App CR` and a related `Catalog` or `HelmRelease` and
+   `HelmRepository`
+   - at this point, Flux or another related deployment operator are already performing Helm Chart deployment
+1. The user has to define an automated version upgrade policy
+   - if it doesn't exist or a user wants to own the object, he creates an `ImageRepository` object, that
+     points to a specific repository (application, helm chart) of an OCI registry. `ImageRepository` discovers
+     and knows all the tags in a remote repository of a registry. There's an exclude regexps list to skip
+     stuff you're sure you won't ever need.
+     - The relationship between `ImageRepository` and `HelmRepository` is very close, we can probably automate
+       `ImageRepository` creation so that users don't have to manage them
```

Comment:
I guess we could automate the creation of these from the targets in the VUC. But then the operator need to know how to parse that info from these CRs, so we cannot work with just any CR. Also if there are multiple targets, nothing really guarantees that they are all the same app actually, so we take the 1st on the list? 🤔 

@ljakimczuk Pointed out that with user creating the ImageRepository + ImagePolicy, with copy pasting users we can end up with applying the latest version of another app. If it fails cos that version does not exists, its okay. The same version existing for that app too is a more scary thing.